### ISSUE TITLE
Check function/subroutine signatures in generated Fortran code

### ISSUE THREAD

**Issue by yguclu** [pyccel/pyccel#200]
It is very important to have unit tests where we verify that we generate the correct function signatures. By this I mean that the following variables' properties must be correct:

- type
- kind
- shape
- intent

To this end we need to parse the generated Fortran files, and extract all information regarding the functions' signatures. @saidctb suggests that we use [LFortran](https://lfortran.org/), which looks very neat and powerful, but we have no experience with that. @krystophny: do you have any recommendation?

**Comment by krystophny** [comment]
@ratnania and I have written a simple parser based on [textX](https://github.com/textX/textX) if you need something more lightweight. You can find it in https://github.com/pyccel/fffi/tree/master/fffi/parser . If anyone is interested in extending its current grammar it should not be too difficult.

**Comment by yguclu** [comment]
@krystophny I see that the `parse` function in `fffi.parser.parser` returns a **textX** abstract syntax tree, so it should be very easy to use in our unit tests. Is maybe the current grammar already sufficient to extract the information I mentioned?

**Comment by krystophny** [comment]
Yes, exactly. The only issue is that I have tweaked this to run examples and tests for fffi as easy as possible, i.e. you can write parse dummy subroutine signatures like

```
subroutine test(a,b)
double precision :: a, b
end
```

This means it doesn't yet support full syntax and all checks for consistency.

**Comment by yguclu** [comment]
If it all boils down to extending `grammar.tx` in `fffi.parser`, we can do that easily. But will this break some of your own tests and examples?

**Comment by krystophny** [comment]
Don't think it will break things, and even if it does, I would adapt to the
more complete grammar. However, as discussed with @yguclu maybe it would be
easier to use LFortran grammar right away for both, fffi and pyccel. It is
BSD licensed so the only remaining question is whether performance is
acceptable compared to a simpler solution.

Am Do., 14. Nov. 2019 um 11:27 Uhr schrieb Yaman Güçlü <
notifications@github.com>:
